pkg/streamingpromql/types/limiting_pool_test.go

```diff
@@ -0,0 +1,208 @@
+
+	// Put a slice back into the pool, the current stat should be updated but peak should be unchanged.
+	p.Put(s100, tracker)
+	require.Equal(t, 2*FPointSize, tracker.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 130*FPointSize, tracker.PeakEstimatedMemoryConsumptionBytes)
+
+	// Get another slice from the pool that doesn't take us over the previous peak.
+	s5, err := p.Get(5, tracker)
+	require.NoError(t, err)
+	require.Equal(t, 8, cap(s5))
+	require.Equal(t, 10*FPointSize, tracker.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 130*FPointSize, tracker.PeakEstimatedMemoryConsumptionBytes)
+
+	// Get another slice from the pool that does take us over the previous peak.
+	s200, err := p.Get(200, tracker)
+	require.NoError(t, err)
+	require.Equal(t, 256, cap(s200))
+	require.Equal(t, 266*FPointSize, tracker.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 266*FPointSize, tracker.PeakEstimatedMemoryConsumptionBytes)
+
+	// Ensure we handle nil slices safely.
+	p.Put(nil, tracker)
+	require.Equal(t, 266*FPointSize, tracker.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 266*FPointSize, tracker.PeakEstimatedMemoryConsumptionBytes)
+
+	assertRejectedQueryCount(t, reg, 0)
+}
+
+func TestLimitingPool_Limited(t *testing.T) {
+	reg, metric := createRejectedMetric()
+	limit := 11 * FPointSize
+	tracker := limiting.NewMemoryConsumptionTracker(limit, metric)
+
+	p := NewLimitingBucketedPool(
+		pool.NewBucketedPool(1, 1000, 2, func(size int) []promql.FPoint { return make([]promql.FPoint, 0, size) }),
+		FPointSize,
+		false,
+	)
+
+	// Get a slice from the pool beneath the limit.
+	s7, err := p.Get(7, tracker)
+	require.NoError(t, err)
+	require.Equal(t, 8, cap(s7))
+	require.Equal(t, 8*FPointSize, tracker.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 8*FPointSize, tracker.PeakEstimatedMemoryConsumptionBytes)
+	assertRejectedQueryCount(t, reg, 0)
+
+	// Get another slice from the pool beneath the limit.
+	s1, err := p.Get(1, tracker)
+	require.NoError(t, err)
+	require.Equal(t, 1, cap(s1))
+	require.Equal(t, 9*FPointSize, tracker.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 9*FPointSize, tracker.PeakEstimatedMemoryConsumptionBytes)
+	assertRejectedQueryCount(t, reg, 0)
+
+	// Return a slice to the pool.
+	p.Put(s1, tracker)
+	require.Equal(t, 8*FPointSize, tracker.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 9*FPointSize, tracker.PeakEstimatedMemoryConsumptionBytes)
+	assertRejectedQueryCount(t, reg, 0)
+
+	// Try to get a slice where the requested size would push us over the limit.
+	_, err = p.Get(4, tracker)
+	expectedError := fmt.Sprintf("the query exceeded the maximum allowed estimated amount of memory consumed by a single query (limit: %d bytes) (err-mimir-max-estimated-memory-consumption-per-query)", limit)
+	require.ErrorContains(t, err, expectedError)
+	require.Equal(t, 8*FPointSize, tracker.CurrentEstimatedMemoryConsumptionBytes)
```

Comment:
should we reach into the pool and check the capacity of the slices here? Even though we reject the Get due to the limit, we still end up performing a Get and Put against the pool so it's possible the slice has grown.

Reply:
Not sure I'm following sorry - are you suggesting we should check that the slice we took from the pool has been returned before we returned the error?

Reply:
Yes, although I'm not entirely sure there's a trivial way to do that.
Additionally it's interesting to confirm that the capacity grew even though we didn't use it.

Reply:
>Yes, although I'm not entirely sure there's a trivial way to do that.

One way to do it would be to introduce an interface for `BucketedPool` so we can mock it out in the test and confirm that the slice is returned, but that will introduce a performance overhead for the production code which I'd prefer to avoid.

The consequences of not returning the slice to the pool are relatively minor, especially given we don't expect to hit the limit too often. 

>Additionally it's interesting to confirm that the capacity grew even though we didn't use it.

Which capacity are you referring to here?

Reply:
`Get` creates a new slice (or takes it from the pool) with a particular capacity, and `Put` puts that slice back in the pool. If something else later needs a slice of the same capacity of the discarded slice, it'll be given that slice, unless GC collects it first.